### PR TITLE
Fix default value for Kokkos_ENABLE_CUDA_LAMBDA

### PR DESCRIPTION
**Pull request by masterleinad**
```
IF (DEFINED CUDA_VERSION AND CUDA_VERSION VERSION_GREATER "7.0")
```
implies that the `Kokkos_ENABLE_CUDA_LAMBDA=ON` should be the default when a `CUDA` version greater than 7.0 is used. Since `CUDA_VERSION` is only defined by `FindCUDA` this variable is in fact only possibly used when configuring with `TriBITS`. On the other hand, we calculate a `CUDA` version and store it as `KOKKOS_COMPILER_CUDA_VERSION` a little bit later. Reordering the `CMake` includes allows using this variable for a more uniform behavior.

### CODE REVIEW

File: cmake/kokkos_enable_options.cmake [hunk 1]
```diff
@@ -52,7 +52,11 @@ KOKKOS_ENABLE_OPTION(EXPLICIT_INSTANTIATION   OFF
   "Whether to explicitly instantiate certain types to lower future compile times")
 SET(KOKKOS_ENABLE_ETI ${KOKKOS_ENABLE_EXPLICIT_INSTANTIATION} CACHE INTERNAL "eti")
 
-IF (DEFINED CUDA_VERSION AND CUDA_VERSION VERSION_GREATER "7.0")
+IF (KOKKOS_ENABLE_CUDA)
+  SET(KOKKOS_COMPILER_CUDA_VERSION "${KOKKOS_VERSION_MAJOR}${KOKKOS_VERSION_MINOR}")
```
Comment (dalg24) [inline cmake/kokkos_enable_options.cmake]: Can't help myself from ranting about this silly naming `KOKKOS_VERSION_{MAJOR,MINOR,PATCH}` that does not refer to the Kokkos version but to the compiler version :@

File: BUILD.md [hunk 2]
```diff
@@ -135,7 +135,7 @@ Options can be enabled by specifiying `-DKokkos_ENABLE_X`.
     * BOOL Default: OFF
 * Kokkos_ENABLE_CUDA_LAMBDA
     * Whether to activate experimental lambda features
-    * BOOL Default: OFF
+    * BOOL Default: ON
```
Comment (dalg24) [inline BUILD.md]: Ping @crtrott

File: cmake/kokkos_enable_options.cmake [hunk 3]
```diff
@@ -52,7 +52,11 @@ KOKKOS_ENABLE_OPTION(EXPLICIT_INSTANTIATION   OFF
...
+ENDIF()
+
+IF (DEFINED KOKKOS_COMPILER_CUDA_VERSION AND KOKKOS_COMPILER_CUDA_VERSION GREATER 70)
   SET(LAMBDA_DEFAULT ON)
```
Comment (crtrott) [inline cmake/kokkos_enable_options.cmake]: Nope, I am against this. As long as NVIDIA doesn't officially support this I would say no. I.e. I want them to make it a non-experimental feature first.

File: cmake/kokkos_tribits.cmake [hunk 4]
```diff
@@ -178,9 +178,9 @@ ENDIF()
 ENDFUNCTION()
 
 MACRO(KOKKOS_SETUP_BUILD_ENVIRONMENT)
+ INCLUDE(${KOKKOS_SRC_PATH}/cmake/kokkos_compiler_id.cmake)
```
Comment (crtrott) [inline cmake/kokkos_tribits.cmake]: Good change. I think a lot of subsequent stuff could depend on the compiler so figuring that out early is good.

File: cmake/kokkos_enable_options.cmake (hunk 1 above)
Comment (crtrott) [reply cmake/kokkos_enable_options.cmake]: wait this is weird where is that coming from? I though KOKKOS_VERSION is actually the version of Kokkos??

**Comment by masterleinad** [comment]
> wait this is weird where is that coming from? I though KOKKOS_VERSION is actually the version of Kokkos??

https://github.com/kokkos/kokkos/blob/c0d6027cbd1fc3c710b9e084a52e92a122e3099f/cmake/kokkos_compiler_id.cmake#L77-L80

`Kokkos_Version` is the version of `Kokkos`.

File: cmake/kokkos_enable_options.cmake (hunk 3 above)
Comment (masterleinad) [reply cmake/kokkos_enable_options.cmake]: OK, I will change that.

**Comment by masterleinad** [comment]
I restored the default to be "OFF" when not using `TriBITS` and also took the liberty of renaming (the internal variable) `KOKKOS_VERSION_*` to `KOKKOS_COMPILER_VERSION_*`.

File: cmake/kokkos_compiler_id.cmake [hunk 5]
```diff
@@ -75,6 +75,6 @@ ELSEIF(KOKKOS_CXX_COMPILER_ID STREQUAL PGI)
 ENDIF()
 
 STRING(REPLACE "." ";" VERSION_LIST ${KOKKOS_CXX_COMPILER_VERSION})
-LIST(GET VERSION_LIST 0 KOKKOS_VERSION_MAJOR)
-LIST(GET VERSION_LIST 1 KOKKOS_VERSION_MINOR)
-LIST(GET VERSION_LIST 2 KOKKOS_VERSION_PATCH)
+LIST(GET VERSION_LIST 0 KOKKOS_COMPILER_VERSION_MAJOR)
+LIST(GET VERSION_LIST 1 KOKKOS_COMPILER_VERSION_MINOR)
+LIST(GET VERSION_LIST 2 KOKKOS_COMPILER_VERSION_PATCH)
```
Comment (dalg24) [inline cmake/kokkos_compiler_id.cmake]: Thank you

Comment (jjwilke) [reply cmake/kokkos_compiler_id.cmake]: Yeah, I poorly named these. Thank you.

**Pull request merged**